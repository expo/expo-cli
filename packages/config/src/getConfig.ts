import JsonFile from '@expo/json-file';
import { formatExecError } from 'jest-message-util';
import path from 'path';

import { spawnSync } from 'child_process';
import { ConfigContext, ExpoConfig } from './Config.types';
import { ConfigError, errorFromJSON } from './Errors';
import { fileExists } from './Modules';
import { serializeAndEvaluate } from './Serialize';

// support all common config types
export const allowedConfigFileNames: string[] = (() => {
  const prefix = 'app';
  return [
    // order is important
    `${prefix}.config.ts`,
    `${prefix}.config.js`,
    `${prefix}.config.json`,
  ];
})();

function isMissingFileCode(code: string): boolean {
  return ['ENOENT', 'MODULE_NOT_FOUND', 'ENOTDIR'].includes(code);
}

function reduceExpoObject(config?: ExpoConfig): ExpoConfig | null {
  if (!config) return null;

  if (typeof config.expo === 'object') {
    // TODO: We should warn users in the future that if there are more values than "expo", those values outside of "expo" will be omitted in favor of the "expo" object.
    return config.expo as ExpoConfig;
  }
  return config;
}

export function findAndEvalConfig(request: ConfigContext): ExpoConfig | null {
  // TODO(Bacon): Support custom config path with `findConfigFile`
  // TODO(Bacon): Should we support `expo` or `app` field with an object in the `package.json` too?

  function testFileName(configFilePath: string): null | Partial<ExpoConfig> {
    if (!fileExists(configFilePath)) return null;

    try {
      return evalConfig(configFilePath, request);
    } catch (error) {
      // If the file doesn't exist then we should skip it and continue searching.
      if (!isMissingFileCode(error.code)) {
        throw error;
      }
    }
    return null;
  }

  if (request.configPath) {
    const config = testFileName(request.configPath);
    if (config) {
      return reduceExpoObject(serializeAndEvaluate(config));
    } else {
      throw new ConfigError(
        `Config with custom path ${request.configPath} couldn't be parsed.`,
        'INVALID_CONFIG'
      );
    }
  }

  for (const configFileName of allowedConfigFileNames) {
    const configFilePath = path.resolve(request.projectRoot, configFileName);
    const config = testFileName(configFilePath);
    if (config) return reduceExpoObject(serializeAndEvaluate(config));
  }

  return null;
}

// We cannot use async config resolution right now because Next.js doesn't support async configs.
// If they don't add support for async Webpack configs then we may need to pull support for Next.js.
function evalConfig(configFile: string, request: ConfigContext): Partial<ExpoConfig> {
  if (configFile.endsWith('.json')) {
    return JsonFile.read(configFile, { json5: true });
  } else {
    try {
      const spawnResults = spawnSync(
        'node',
        [
          require.resolve('@expo/config/build/scripts/read-config.js'),
          '--colors',
          configFile,
          JSON.stringify({ ...request, config: serializeAndEvaluate(request.config) }),
        ],
        {}
      );

      if (spawnResults.status === 0) {
        const spawnResultString = spawnResults.stdout.toString('utf8').trim();
        const logs = spawnResultString.split('\n');
        // Get the last console log to prevent parsing anything logged in the config.
        const lastLog = logs.pop()!;
        for (const log of logs) {
          // Log out the logs from the config
          console.log(log);
        }
        // Parse the final log of the script, it's the serialized config
        return JSON.parse(lastLog);
      } else {
        // Parse the error data and throw it as expected
        const errorData = JSON.parse(spawnResults.stderr.toString('utf8'));
        throw errorFromJSON(errorData);
      }
    } catch (error) {
      if (isMissingFileCode(error.code) || !(error instanceof SyntaxError)) {
        throw error;
      }
      const message = formatExecError(
        error,
        { rootDir: request.projectRoot, testMatch: [] },
        { noStackTrace: true },
        undefined,
        true
      );
      throw new ConfigError(`\n${message}`, 'INVALID_CONFIG');
    }
  }
}
