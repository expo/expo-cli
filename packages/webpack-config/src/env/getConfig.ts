import { ExpoConfig, getConfigForPWA } from '@expo/config';

import { Environment } from '../types';
import { getPaths } from './paths';

/**
 * Get the Expo project config in a way that's optimized for web.
 *
 * @param env Environment properties used for getting the Expo project config.
 * @category env
 */
function getConfig(env: Pick<Environment, 'projectRoot' | 'config' | 'locations'>): ExpoConfig {
  if (env.config) {
    return env.config;
  }
  const locations = env.locations || getPaths(env.projectRoot);
  // Fill all config values with PWA defaults
  return getConfigForPWA(env.projectRoot, locations.absolute, {
    templateIcon: locations.template.get('icon.png'),
  });
}

export default getConfig;
