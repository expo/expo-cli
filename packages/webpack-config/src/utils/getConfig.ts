import { ExpoConfig, getConfigForPWA } from '@expo/config';

import { Environment } from '../types';
import { getPaths } from './paths';

function getConfig(env: Environment): ExpoConfig {
  if (env.config) {
    return env.config;
  }
  const locations = env.locations || getPaths(env.projectRoot);
  // Fill all config values with PWA defaults
  return getConfigForPWA(env.projectRoot, locations.absolute, {
    templateIcon: locations.template.get('icon.png'),
  });
}

export default getConfig;
