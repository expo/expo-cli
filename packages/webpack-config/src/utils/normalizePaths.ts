/**
 * Converts absolute paths to relative paths for testing purposes.
 *
 * @param initial
 * @param transformString
 * @internal
 */
export default function normalizePaths(
  initial: any,
  transformString: (value: string) => string
): any {
  if (initial == null) {
    return initial;
  } else if (typeof initial === 'string') {
    return transformString(initial);
  } else if (Array.isArray(initial)) {
    return initial.map(value => normalizePaths(value, transformString));
  } else if (typeof initial === 'object') {
    let result: { [key: string]: any } = {};
    for (const prop of Object.keys(initial)) {
      result[prop] = normalizePaths(initial[prop], transformString);
    }
    return result;
  } else {
    return initial;
  }
}
