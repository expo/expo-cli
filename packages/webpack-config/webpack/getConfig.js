const { getConfigForPWA } = require('@expo/config');
const getLocations = require('./webpackLocations');

module.exports = function(env) {
  const locations = getLocations(env.projectRoot);
  // Fill all config values with PWA defaults
  return getConfigForPWA(env.projectRoot, locations.absolute, {
    templateIcon: locations.template.get('icon.png'),
  });
};
