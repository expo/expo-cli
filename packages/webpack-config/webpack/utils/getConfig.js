const { getConfigForPWA } = require('@expo/config');
const getPaths = require('./getPaths');

module.exports = function(env) {
  if (env.config) {
    return env.config;
  }
  const locations = getPaths(env);
  // Fill all config values with PWA defaults
  return getConfigForPWA(env.projectRoot, locations.absolute, {
    templateIcon: locations.template.get('icon.png'),
  });
};
