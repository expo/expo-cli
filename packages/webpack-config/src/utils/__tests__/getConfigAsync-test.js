import path from 'path';
import getConfig from '../getConfig';
import normalizePaths from '../normalizePaths';

const projectRoot = path.resolve(__dirname, '../../../tests/basic');
const mode = 'development';
const env = { projectRoot, mode };

it(`has consistent defaults`, () => {
  const config = getConfig(env);
  const normalized = normalizePaths(config, value => value.split('packages/webpack-config/').pop());
  expect(normalized).toMatchSnapshot();
});
