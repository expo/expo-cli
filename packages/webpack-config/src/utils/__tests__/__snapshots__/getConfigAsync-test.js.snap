// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has consistent defaults 1`] = `
Object {
  "android": Object {},
  "assetBundlePatterns": Array [
    "**/*",
  ],
  "description": "A Neat Expo App",
  "icon": "./assets/icon.png",
  "ios": Object {
    "supportsTablet": true,
  },
  "name": "basic",
  "orientation": "portrait",
  "platforms": Array [
    "ios",
    "android",
    "web",
  ],
  "primaryColor": "#4630EB",
  "privacy": "public",
  "sdkVersion": "33.0.0",
  "slug": "basic",
  "splash": Object {
    "backgroundColor": "#ffffff",
    "image": "./assets/splash.png",
    "resizeMode": "contain",
  },
  "updates": Object {
    "fallbackToCacheTimeout": 0,
  },
  "version": "1.0.0",
  "web": Object {
    "backgroundColor": "#ffffff",
    "barStyle": "default",
    "build": Object {
      "output": "web-build",
      "publicPath": "/",
      "rootId": "root",
    },
    "crossorigin": undefined,
    "dangerous": Object {
      "noJavaScriptMessage": "Oh no! It looks like JavaScript is not enabled in your browser.",
    },
    "description": "A Neat Expo App",
    "dir": "auto",
    "display": "standalone",
    "icons": Array [
      Object {
        "destination": "apple/icons",
        "size": Array [
          192,
          512,
        ],
        "src": "tests/basic/assets/icon.png",
      },
      Object {
        "destination": "apple/icons",
        "ios": true,
        "sizes": 180,
        "src": "tests/basic/assets/icon.png",
      },
    ],
    "lang": "en",
    "meta": Object {
      "apple": Object {
        "barStyle": "default",
        "formatDetection": "telephone=no",
        "mobileWebAppCapable": "yes",
        "touchFullscreen": "yes",
      },
      "viewport": "width=device-width,initial-scale=1,minimum-scale=1,maximum-scale=1.00001,viewport-fit=cover",
    },
    "name": "basic",
    "orientation": "portrait",
    "preferRelatedApplications": true,
    "relatedApplications": Array [],
    "scope": undefined,
    "shortName": "basic",
    "startUrl": ".",
    "startupImages": Array [
      Object {
        "color": "#ffffff",
        "destination": "apple/splash",
        "orientation": "portrait",
        "resizeMode": "contain",
        "src": "tests/basic/assets/splash.png",
        "supportsTablet": true,
      },
    ],
    "themeColor": "#4630EB",
  },
}
`;
